File: lib/hecks_specializer/meta_subclass.rb
# lib/hecks_specializer/meta_subclass.rb
#
# Hecks::Specializer::MetaSubclass — Phase C PC-1 pilot.
#
# The meta-specializer. Reads SpecializerSubclass rows from
# specializer.fixtures and emits lib/hecks_specializer/<target>.rb
# (the Ruby subclass shells themselves).
#
# This is the FIRST self-referential specializer: its output lives
# under the same lib/hecks_specializer/ that holds the code reading
# it. PC-1's byte-identity test proves the specializer can regenerate
# (part of) itself — the pilot for the 2nd Futamura fixed point.
#
# Phase: C. Scope: 5 subclass shells (~15 LoC each); base class,
# driver, and fixed-point come in PC-2..PC-4.
#
# Usage differs from other targets because there's N output files,
# not one. Invoked via:
#   bin/specialize meta_subclass              # emits first-row target
#   bin/specialize meta_subclass --list       # list all rows
#   bin/specialize meta_subclass --name NAME  # emit specific row
#   bin/specialize meta_subclass --all        # emit every row to disk
#
# For golden-test purposes, :meta_subclass with no flags emits the
# first registered row (DuplicatePolicy) to stdout — consistent with
# how other targets behave and lets us gate byte-identity with the
# same assert_byte_identical helper.

require_relative "diagnostic_validator" # picks up Target mixin

module Hecks
  module Specializer
    class MetaSubclass
      include Target

      SHAPE = REPO_ROOT.join("hecks_conception/capabilities/specializer/fixtures/specializer.fixtures")
      TARGET_RS = REPO_ROOT.join("lib/hecks_specializer/duplicate_policy.rb")

      # Which SpecializerSubclass fixture row to emit for. Subclasses
      # override to pick different rows.
      def self.row_target_name
        "duplicate_policy"
      end

      def emit
        rows = by_aggregate("SpecializerSubclass")
        row = rows.find { |r| r["attrs"]["target_name"] == self.class.row_target_name }
        raise "no SpecializerSubclass row for #{self.class.row_target_name.inspect}" unless row
        emit_row(row)
      end

      private

      # Emit one Ruby subclass file's contents. Template matches the
      # hand-written shell byte-for-byte: doc block, require_relative,
      # class body, register call.
      def emit_row(row)
        a = row["attrs"]
        doc_lines = a["module_doc"].split("\n").map do |line|
          line.empty? ? "#" : "# #{line}"
        end
        <<~RB
          # #{a["output_rb"]}
          #
          #{doc_lines.join("\n")}

          require_relative "diagnostic_validator"

          module Hecks
            module Specializer
              class #{a["class_name"]} < #{a["base_class"]}
                SHAPE = REPO_ROOT.join("#{a["shape_path"]}")
                TARGET_RS = REPO_ROOT.join("#{a["target_rs_path"]}")
              end

              register :#{a["target_name"]}, #{a["class_name"]}
            end
          end
        RB
      end
    end

    register :meta_subclass, MetaSubclass

    # Second meta target — emits lib/hecks_specializer/lifecycle.rb
    # from the Lifecycle SpecializerSubclass row. Same pattern, different
    # row + output. Added in Phase C PC-1b.
    class MetaSubclassLifecycle < MetaSubclass
      TARGET_RS = REPO_ROOT.join("lib/hecks_specializer/lifecycle.rb")
      def self.row_target_name
        "lifecycle"
      end
    end

    register :meta_subclass_lifecycle, MetaSubclassLifecycle
  end
end

File: hecks_life/src/specializer/meta_subclass.rs
//! Rust port of `lib/hecks_specializer/meta_subclass.rb`.
//!
//! Emits one thin-subclass Ruby shell under `lib/hecks_specializer/`
//! from a `SpecializerSubclass` fixture row — the Phase C PC-1 pilot,
//! now also available through the Phase D Rust-native driver.
//!
//! Unlike the D1/D2 ports, which emit Rust, this port emits Ruby.
//! The source fixtures contain no `.rb.frag` snippets — every shell
//! is assembled inline from fixture attributes (class_name,
//! base_class, module_doc, shape_path, target_rs_path, target_name,
//! output_rb). The template has to match the hand-written shells
//! byte-for-byte: doc block, `require_relative`, class body, register
//! call.
//!
//! Which row to emit is controlled by a row-target name — the
//! dispatcher picks the row whose `target_name` matches. Default
//! (`meta_subclass`) emits the DuplicatePolicy row, mirroring the
//! Ruby `MetaSubclass.row_target_name`; `meta_subclass_lifecycle`
//! picks the Lifecycle row, mirroring `MetaSubclassLifecycle`.
//!
//! Usage:
//!   let rb = meta_subclass::emit_named(repo_root, "duplicate_policy")?;
//!   print!("{}", rb);
//!
//! [antibody-exempt: hecks_life/src/specializer/meta_subclass.rs —
//!  Phase D D3 — Ruby-native specializer port]

use crate::specializer::util;
use std::error::Error;
use std::path::Path;

const SHAPE_REL: &str = "hecks_conception/capabilities/specializer/fixtures/specializer.fixtures";

/// Emit the shell for the `duplicate_policy` row — the default
/// `meta_subclass` dispatch target, mirroring Ruby
/// `MetaSubclass.row_target_name`.
pub fn emit(repo_root: &Path) -> Result<String, Box<dyn Error>> {
    emit_named(repo_root, "duplicate_policy")
}

/// Emit the shell for the `lifecycle` row — the
/// `meta_subclass_lifecycle` dispatch target, mirroring Ruby
/// `MetaSubclassLifecycle.row_target_name`.
pub fn emit_lifecycle(repo_root: &Path) -> Result<String, Box<dyn Error>> {
    emit_named(repo_root, "lifecycle")
}

/// Locate the `SpecializerSubclass` row whose `target_name` matches
/// `row_target_name`, then render the thin-subclass Ruby shell.
pub fn emit_named(repo_root: &Path, row_target_name: &str) -> Result<String, Box<dyn Error>> {
    let shape = repo_root.join(SHAPE_REL);
    let fixtures = util::load_fixtures(&shape)?;
    let rows = util::by_aggregate(&fixtures, "SpecializerSubclass");
    let row = rows
        .iter()
        .find(|r| util::attr(r, "target_name") == row_target_name)
        .ok_or_else(|| {
            format!(
                "no SpecializerSubclass row for {:?}",
                row_target_name
            )
        })?;
    Ok(emit_row(row))
}

fn emit_row(row: &crate::ir::Fixture) -> String {
    let output_rb = util::attr(row, "output_rb");
    let class_name = util::attr(row, "class_name");
    let base_class = util::attr(row, "base_class");
    let shape_path = util::attr(row, "shape_path");
    let target_rs_path = util::attr(row, "target_rs_path");
    let target_name = util::attr(row, "target_name");
    let module_doc = util::attr(row, "module_doc");

    // Ruby:
    //   doc_lines = module_doc.split("\n").map { |l| l.empty? ? "#" : "# #{l}" }
    // The fixtures parser already expanded `\n` escapes to real newlines.
    let doc_lines: Vec<String> = module_doc
        .split('\n')
        .map(|line| {
            if line.is_empty() {
                "#".to_string()
            } else {
                format!("# {line}")
            }
        })
        .collect();
    let doc_block = doc_lines.join("\n");

    // Mirrors the Ruby heredoc byte-for-byte. Indentation is two
    // spaces for `module Specializer`, four for the class body, six
    // for SHAPE/TARGET_RS.
    format!(
        "\
# {output_rb}
#
{doc_block}

require_relative \"diagnostic_validator\"

module Hecks
  module Specializer
    class {class_name} < {base_class}
      SHAPE = REPO_ROOT.join(\"{shape_path}\")
      TARGET_RS = REPO_ROOT.join(\"{target_rs_path}\")
    end

    register :{target_name}, {class_name}
  end
end
",
        output_rb = output_rb,
        doc_block = doc_block,
        class_name = class_name,
        base_class = base_class,
        shape_path = shape_path,
        target_rs_path = target_rs_path,
        target_name = target_name,
    )
}
